include/zephyr/kernel.h

```diff
@@ -961,6 +961,14 @@ int k_thread_cpu_pin(k_tid_t thread, int cpu);
  * will be canceled.  On resume, the thread will begin running
  * immediately and return from the blocked call.
  *
+ * Care must taken when an application issues either k_thread_suspend() or
+ * k_thread_abort() on an SMP platform. While threads may block when the target
+ * thread is active on another CPU, ISRs will spin. This means that on SMP
+ * platforms, it is possible for application code to create a deadlock
+ * condition by simultaneously suspending or aborting using at least one
+ * suspension or termination from interrupt context. Zephyr cannot detect all
+ * such conditions.
```

Comment:
> The old k_thread_abort() was deadlock-proof.

No, the old k_thread_abort() was NOT deadlock proof, and the current documentation explicitly states it. 

One such scenario wherein a deadlock can occur is if both CPU0 and CPU1 are in ISRs and they each try to abort the other CPU's interrupted thread. As both threads are currently active as they were interrupted, the _THREAD_ABORTING bit is set for both.  The next step in this sequence is for both of these CPUs to confirm that the thread they want to abort is marked as _THREAD_ABORTING and that it is not its current/interrupted thread. Since they are both in an ISR, they will each spin waiting for the thread to be aborted--deadlocked. Is this deadlock likely? No. But can it occur? Yes.

As the new proposed k_thread_suspend() shares the same infrastructure as k_thread_abort(), it too is vulnerable to deadlock. Thus, it must be mentioned.

I do have some ideas on how this unlikely deadlock can be prevented, but those are future plans.

Reply:
No, this is handled.  Or should be; "was designed to be", anyway.  I'm not going to hold this code up as a model of perfection or anything, but I will say that this (the second attempt to make abort SMP-safe) was excruciatingly hard to get right the first time, which informs my conservatism about this review.

There is a check inside the scheduler lock at the top of the abort path: the first abort-in-a-cycle thread to make it through the abort path with set the aborting flag and spin, the second will see the flag set and synchronously abort the current thread before spinning.  That guarantees that both interrupts exit k_thread_abort() in finite time, and the lock guarantees serialization of changes to the aborting flag such that if one CPU sets it the other is guaranteed not to have already check it.

It's possible there's a bug I'm not seeing or a glitch I'm forgetting.  But it's not that.

Reply:
I don't follow?  Specifically I don't see how you get to step 11 ("CPU1 ISR determines thread0 is not aborting"): that thread will have been affirmatively flagged aborting before the lock is released (step 10).

Really, can we just remove this comment?  I'm almost certain it's not correct that we have an unresolvable deadlock in k_sched_abort().  And even if it is correct, we should file that as a bug, write a reproducer and commit to fixing it instead of trying to document that kind of misfeature as a "intended undefined behavior" kind of thing.  Other OS's get this right, we should too.